packages/@aws-cdk/aws-ec2/lib/cfn-init.ts

```diff
@@ -97,7 +97,9 @@ export class CloudFormationInit {
     // Note: This will not reflect mutations made after attaching.
     const bindResult = this.bind(attachedResource.stack, attachOptions);
     attachedResource.addMetadata('AWS::CloudFormation::Init', bindResult.configData);
-    const fingerprint = contentHash(JSON.stringify(bindResult.configData)).substr(0, 16);
+    // Need to resolve the various tokens from assets in the config so the fingerprint is accurate.
+    const resolvedConfig = attachedResource.stack.resolve(bindResult.configData);
```

Comment:
I hinted at this not being good enough in our chat conversation, but I didn't actually send you a link on this.

This solution will work *for now*, but it's going to run afoul of the same limitation as this issue: https://github.com/aws/aws-cdk/issues/10136

Reply:
Added an `assetHash` to the `InitElementConfig` type and am now propagating that up back to the fingerprint.